crates/resolver-tests/tests/pubgrub.rs

```diff
@@ -0,0 +1,453 @@
+}
+
+#[test]
+fn test_o_no_implicit_feature() {
+    let reg = registry(vec![
+        pkg("c"),
+        pkg_dep_with("ureq", vec!["c".opt()], &[("cookies", &["dep:c"])]),
+        pkg_dep_with("dep", vec![dep("ureq")], &[("cookies", &["ureq/c"])]),
+    ]);
+
+    let deps = vec!["dep".with(&["cookies"])];
+    let mut sat_resolver = SatResolver::new(&reg);
+    assert!(resolve_and_validated(deps, &reg, &mut sat_resolver).is_err());
+}
+
+#[test]
+fn test_optional_self_dependency() {
+    let reg = registry(vec![pkg_dep("a", vec!["a".opt()])]);
+
+    let deps = vec!["a".with(&["a"])];
+
+    // Cyclic dependencies are not checked in the SAT resolver
+    assert!(resolve(deps.clone(), &reg).is_err());
+    assert!(SatResolver::new(&reg).sat_resolve(&deps));
+}
+
+#[test]
+fn test_p_weak_dependencies() {
+    let reg = registry(vec![
+        pkg_dep("a", vec![dep("bad")]),
+        pkg_dep_with("b", vec!["a".opt()], &[("perf-literal", &["dep:a"])]),
+        pkg_dep_with(
+            "c",
+            vec!["b".opt()],
+            &[
+                ("perf-literal", &["b?/perf-literal"]),
+                ("perf-literal-multisubstring", &["dep:b"]),
+            ],
+        ),
+        pkg_dep_with("dep", vec![dep("c")], &[("default", &["c/perf-literal"])]),
+    ]);
+
+    let deps = vec!["dep".with(&["default"])];
+
+    // Weak dependencies are not supported yet in the dependency resolver
+    assert!(resolve(deps.clone(), &reg).is_err());
+    assert!(SatResolver::new(&reg).sat_resolve(&deps));
+}
+
+#[test]
+fn test_q_features_propagation() {
+    let reg = registry(vec![
+        pkg_dep_with("proc-macro2", vec![], &[("proc-macro", &[])]),
+        pkg_dep_with(
+            "syn",
+            vec![dep("proc-macro2")],
+            &[("proc-macro", &["proc-macro2/proc-macro"])],
+        ),
+    ]);
+
+    let deps = vec!["syn".with(&["proc-macro"])];
+    let mut sat_resolver = SatResolver::new(&reg);
+    assert!(resolve_and_validated(deps, &reg, &mut sat_resolver).is_ok());
+}
+
+#[test]
+fn test_r_features_propagation_and_unification() {
+    let reg = registry(vec![
+        pkg_dep_with("quote", vec![], &[("proc-macro", &[])]),
+        pkg_dep_with(
+            "syn",
+            vec!["quote".opt()],
+            &[("default", &["quote", "quote/proc-macro"])],
+        ),
+    ]);
+
+    let deps = vec!["syn".with(&["default"])];
+    let mut sat_resolver = SatResolver::new(&reg);
+    assert!(resolve_and_validated(deps, &reg, &mut sat_resolver).is_ok());
+}
+
+#[test]
+fn test_s_feature_shadowing_bug() {
+    let reg = registry(vec![
+        pkg_dep_with(
+            "a",
+            vec!["b".opt(), "x".opt()],
+            &[("b", &["x"]), ("default", &["b/native"])],
+        ),
+        pkg_dep_with("b", vec![], &[("native", &[])]),
+    ]);
+
+    let deps = vec!["a".with(&["default"])];
+
+    // Bug in the dependency resolver: feature `b` of package `a` should not be activated by `b/native`
```

Comment:
I think this is a bug in the dependency resolver. Has an issue already been created for this ?

Reply:
I thought it was a bug when I first came across it as well. But it is intentional documented behavior: https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Odd.20cargo.20resolver.20behavior